[docs-only] Update testing docs (#6346)
owncloud/ocis: docs/ocis/development/testing.md

{{< /hint >}}

```bash
BEHAT_FEATURE='tests/acceptance/features/apiGraph/createUser.feature:26' \
make -C tests/acceptance/docker test-ocis-feature-ocis-storage
```

Similarly, with S3 storage;

```bash
# run a whole feature
BEHAT_FEATURE='tests/acceptance/features/apiGraph/createUser.feature' \
make -C tests/acceptance/docker test-ocis-feature-s3ng-storage

# run a single scenario
BEHAT_FEATURE='tests/acceptance/features/apiGraph/createUser.feature:26' \
make -C tests/acceptance/docker test-ocis-feature-s3ng-storage
```

In the same way, tests transferred from ownCloud core can be run as:

```bash
# run a whole feature
BEHAT_FEATURE='tests/acceptance/features/coreApiAuth/webDavAuth.feature' \
make -C tests/acceptance/docker test-core-feature-ocis-storage

# run a single scenario
BEHAT_FEATURE='tests/acceptance/features/coreApiAuth/webDavAuth.feature:13' \
make -C tests/acceptance/docker test-core-feature-ocis-storage
```

{{< hint info >}}

> The tests suites transferred from ownCloud core have `coreApi` prefixed

Reviewer: ```suggestion
The test suites transferred from ownCloud core have `coreApi` prefixed
```

For example, command:

```bash
make -C tests/acceptance/docker localApiTests-apiGraph-ocis
```

runs the same tests as the `localApiTests-apiGraph-ocis` CI pipeline, which runs the oCIS test suite "apiGraph" against the oCIS server with ocis storage.

And command:

```bash
make -C tests/acceptance/docker localApiTests-apiGraph-s3ng
```

runs the oCIS test suite `apiGraph` against the oCIS server with s3 storage.

{{< hint info >}}
While running the tests, oCIS server is started with [ociswrapper](https://github.com/owncloud/ocis/blob/master/tests/ociswrapper/README.md) (i.e. `WITH_WRAPPER=true`) by default. In order to run the tests without ociswrapper, provide `WITH_WRAPPER=false` when running the tests. For example:

```bash
WITH_WRAPPER=false \
BEHAT_FEATURE='tests/acceptance/features/apiGraph/createUser.feature:26' \
make -C tests/acceptance/docker test-ocis-feature-ocis-storage
```

But some test suites that are tagged with `@env-config` require the oCIS server to be run with ociswrapper. So, running those tests require `WITH_WRAPPER=true` (default setting).
{{< /hint >}}

#### Tests transferred from ownCloud core (prefix `coreApi`)

> Command `make -C tests/acceptance/docker Core-API-Tests-ocis-storage-3` runs the same tests as the `Core-API-Tests-ocis-storage-3` CI pipeline, which runs the third (out of ten) test suites transferred from the ownCloud core against the oCIS server with ocis storage.

Reviewer: ```suggestion
Command `make -C tests/acceptance/docker Core-API-Tests-ocis-storage-3` runs the same tests as the `Core-API-Tests-ocis-storage-3` CI pipeline, which runs the third (out of ten) test suite groups transferred from ownCloud core against the oCIS server with ocis storage.
```

There might be 28 test suites, and the CI is split into 10 pipelines. So each pipeline runs a "group" of 2 or 3 test suites.

> And `make -C tests/acceptance/docker Core-API-Tests-s3ng-storage-3` runs the third (out of ten) test suite transferred from the ownCloud core against the oCIS server with s3 storage.

Reviewer: ```suggestion
And `make -C tests/acceptance/docker Core-API-Tests-s3ng-storage-3` runs the third (out of ten) test suite groups transferred from ownCloud core against the oCIS server with s3 storage.
```

> ### Run full test suite

Reviewer 2: ```suggestion
### Run Full Test Suite
```

> #### Local oCIS tests (prefix `api`)

Reviewer 2: ```suggestion
#### Local oCIS Tests (prefix `api`)
```
Pls use [headlinecapitalisation.com](https://headlinecapitalization.com) for proper writing of section headlines.

For example, command:

```bash
make -C tests/acceptance/docker localApiTests-apiGraph-ocis
```

runs the same tests as the `localApiTests-apiGraph-ocis` CI pipeline, which runs the oCIS test suite "apiGraph" against the oCIS server with ocis storage.

And command:

```bash
make -C tests/acceptance/docker localApiTests-apiGraph-s3ng
```

runs the oCIS test suite `apiGraph` against the oCIS server with s3 storage.

{{< hint info >}}
While running the tests, oCIS server is started with [ociswrapper](https://github.com/owncloud/ocis/blob/master/tests/ociswrapper/README.md) (i.e. `WITH_WRAPPER=true`) by default. In order to run the tests without ociswrapper, provide `WITH_WRAPPER=false` when running the tests. For example:

```bash
WITH_WRAPPER=false \
BEHAT_FEATURE='tests/acceptance/features/apiGraph/createUser.feature:26' \
make -C tests/acceptance/docker test-ocis-feature-ocis-storage
```

But some test suites that are tagged with `@env-config` require the oCIS server to be run with ociswrapper. So, running those tests require `WITH_WRAPPER=true` (default setting).
{{< /hint >}}

> #### Tests transferred from ownCloud core (prefix `coreApi`)

Reviewer 2: ```suggestion
#### Tests Transferred from ownCloud core (prefix `coreApi`)
```

Command `make -C tests/acceptance/docker Core-API-Tests-ocis-storage-3` runs the same tests as the `Core-API-Tests-ocis-storage-3` CI pipeline, which runs the third (out of ten) test suite groups transferred from ownCloud core against the oCIS server with ocis storage.

> And `make -C tests/acceptance/docker Core-API-Tests-s3ng-storage-3` runs the third (out of ten) test suite groups transferred from ownCloud core against the oCIS server with s3 storage.
> 
> ### Run single feature test

Reviewer 2: ```suggestion
And `make -C tests/acceptance/docker Core-API-Tests-s3ng-storage-3` runs the third (out of ten) test suite groups transferred from ownCloud core against the oCIS server with s3 storage.

### Run Single Feature Test
```

The tests for a single feature (a feature file) can also be run against the different storage backends. To do that, multiple make targets with the schema **test-_\<test-source\>_-feature-_\<storage-backend\>_** are available. To select a single feature you have to add an additional `BEHAT_FEATURE=<path-to-feature-file>` parameter when invoking the make command.

For example;

```bash
BEHAT_FEATURE='tests/acceptance/features/apiGraph/createUser.feature' \
make -C tests/acceptance/docker test-ocis-feature-ocis-storage
```

{{< hint info >}}
`BEHAT_FEATURE` must be pointing to a valid feature file
{{< /hint >}}

And to run a single scenario in a feature, you can do:

{{< hint info >}}

> A specific scenario from a feature can be run by adding `:<line-number>` at the end of the feature file path. For example, to run the scenario at line 26 of the feature file `apiGraph/createUser.feature`, simply add the line number like this: `apiGraph/createUser.feature:26`

Reviewer 3: I meant something like this
```suggestion
A specific scenario from a feature can be run by adding `:<line-number>` at the end of the feature file path. For example, to run the scenario at line 26 of the feature file `apiGraph/createUser.feature`, simply add the line number like this: `apiGraph/createUser.feature:26` . Note that the line numbers mentioned in the examples might not always point to a scenario, so always check the line numbers before running the test.
```

# run a single scenario
BEHAT_FEATURE='tests/acceptance/features/apiGraph/createUser.feature:26' \
make -C tests/acceptance/docker test-ocis-feature-s3ng-storage
```

In the same way, tests transferred from ownCloud core can be run as:

```bash
# run a whole feature
BEHAT_FEATURE='tests/acceptance/features/coreApiAuth/webDavAuth.feature' \
make -C tests/acceptance/docker test-core-feature-ocis-storage

# run a single scenario
BEHAT_FEATURE='tests/acceptance/features/coreApiAuth/webDavAuth.feature:13' \
make -C tests/acceptance/docker test-core-feature-ocis-storage
```

{{< hint info >}}
The tests suites transferred from ownCloud core have `coreApi` prefixed
{{< /hint >}}

### oCIS image to be tested (skip local image build)

By default, the tests will be run against the docker image built from your current working state of the oCIS repository. For some purposes it might also be handy to use an oCIS image from Docker Hub. Therefore, you can provide the optional flag `OCIS_IMAGE_TAG=...` which must contain an available docker tag of the [owncloud/ocis registry on Docker Hub](https://hub.docker.com/r/owncloud/ocis) (e.g. 'latest').

```bash
OCIS_IMAGE_TAG=latest \

> make -C tests/acceptance/docker localApiTests-apiGraph-ocis
> ```
> 
> ### Test log output

Reviewer 2: ```suggestion
make -C tests/acceptance/docker localApiTests-apiGraph-ocis
```

### Test Log Output
```

During testing we start a redis and oCIS docker container. These will not be stopped automatically. You can stop them with:

> ```bash
> make -C tests/acceptance/docker clean
> ```
> 
> ## Testing with test suite natively installed

Reviewer 2: ```suggestion
```bash
make -C tests/acceptance/docker clean
```

## Testing With Test Suite Natively Installed
```

> ## Testing with test suite natively installed

Reviewer 2: ```suggestion
## Testing With Test Suite Natively Installed
```

`PROXY_ENABLE_BASIC_AUTH` will allow the acceptance tests to make requests against the provisioning api (and other endpoints) using basic auth.

> #### Run local oCIS tests (prefix `api`)

Reviewer 2: ```suggestion
#### Run Local oCIS Tests (prefix `api`)
```

```bash
make test-acceptance-api \
TEST_SERVER_URL=https://localhost:9200 \
TEST_WITH_GRAPH_API=true \
TEST_OCIS=true \
```

> #### Run tests transferred from ownCloud core (prefix `coreApi`)

Reviewer 2: ```suggestion
#### Run Tests Transferred From ownCloud core (prefix `coreApi`)
```

```bash
make test-acceptance-from-core-api \
TEST_SERVER_URL=https://localhost:9200 \
TEST_WITH_GRAPH_API=true \
TEST_OCIS=true \
```

Useful environment variables:

`TEST_SERVER_URL`: oCIS server url. Please, adjust the server url according to your setup.

`BEHAT_FEATURE`: to run a single feature

{{< hint info >}}
A specific scenario from a feature can be run by adding `:<line-number>` at the end of the feature file path. For example, to run the scenario at line 26 of the feature file `apiGraph/createUser.feature`, simply add the line number like this: `apiGraph/createUser.feature:26`. Note that the line numbers mentioned in the examples might not always point to a scenario, so always check the line numbers before running the test.
{{< /hint >}}

> Example:
>
> BEHAT_FEATURE=tests/acceptance/features/apiGraph/createUser.feature
>
> Or
>
> BEHAT_FEATURE=tests/acceptance/features/apiGraph/createUser.feature:12

`BEHAT_SUITE`: to run a single suite

> Example:
>
> BEHAT_SUITE=apiGraph

`STORAGE_DRIVER`: to run tests with a different user storage driver. Available options are `ocis` (default), `owncloudsql` and `s3ng`

> Example:
>
> STORAGE_DRIVER=owncloudsql

> ### Use existing tests for BDD

Reviewer 2: ```suggestion
### Use Existing Tests For BDD
```

1. locally run each of the tests marked with that issue in the expected failures file.

   E.g.:

   ```bash
   make test-acceptance-from-core-api \

>    TEST_SERVER_URL=https://localhost:9200 \

Reviewer 3: Should we do it something like this or hard coding is fine?
```suggestion
   TEST_SERVER_URL=<ocis-instance-url> \
```

Author: Most people run it locally so it should be fine but give me suggestions, I couldn't decide here

Author: Let it be as an example. I will write a note on `TEST_SERVER_URL` envvar

Author: Added here https://github.com/owncloud/ocis/blob/82e1423056d8d214a57b57a5ae2285561659a8e6/docs/ocis/development/testing.md?plain=1#L95-L97

   TEST_OCIS=true \
   TEST_WITH_GRAPH_API=true \
   STORAGE_DRIVER=OCIS \

>    BEHAT_FEATURE='tests/acceptance/features/coreApiVersions/fileVersions.feature:147'

Reviewer 3: maybe we can also add a little bit of a hint or note saying that the line number might change because  
this might not work if we change something in feature files `tests/acceptance/features/coreApiVersions/fileVersions.feature:147` 

OR
`tests/acceptance/features/coreApiVersions/fileVersions.feature:<line-number-pointing-to-a-scenario>`

Author: Let it be as an example. I will write a note on running specific scenario using line number

Author: added here https://github.com/owncloud/ocis/blob/82e1423056d8d214a57b57a5ae2285561659a8e6/docs/ocis/development/testing.md?plain=1#L205

2. Use `OcisConfigHelper.php` for helper functions - provides functions to reconfigure the running oCIS instance.
3. Recommended: add the new step implementations in `OcisConfigContext.php`

> ## Running test suite with email service (@email)

Reviewer 2: ```suggestion
## Running Test Suite With Email Service (@email)
```

Test suites that are tagged with `@email` require an email service. We use inbucket as the email service in our tests.

> ### Setup inbucket

Reviewer 2: ```suggestion
### Setup Inbucket
```

Run the following command to setup inbucket

```bash
docker run -d -p9000:9000 -p2500:2500 --name inbucket inbucket/inbucket
```

### Run oCIS

Documentation for environment variables is available [here](https://owncloud.dev/services/notifications/#environment-variables)

```bash
# init oCIS
IDM_ADMIN_PASSWORD=admin \
ocis/bin/ocis init --insecure true

# run oCIS
PROXY_ENABLE_BASIC_AUTH=true \
NOTIFICATIONS_SMTP_HOST=localhost \
NOTIFICATIONS_SMTP_PORT=2500 \
NOTIFICATIONS_SMTP_INSECURE=true \
ocis/bin/ocis server
```

### Run the acceptance test

Run the acceptance test with the following command:

```bash
TEST_WITH_GRAPH_API=true \
TEST_OCIS=true \
TEST_SERVER_URL="https://localhost:9200" \
EMAIL_HOST="localhost" \
EMAIL_PORT=9000 \
BEHAT_FEATURE="tests/acceptance/features/apiEmailNotification/emailNotification.feature" \
make test-acceptance-api
```

> ## Running tests for parallel deployment
> 
> ### Setup the parallel deployment environment

Reviewer 2: ```suggestion
## Running Tests for Parallel Deployment

### Setup the Parallel Deployment Environment
```
